File: Ps.6/sentimental-credit/credit.py


def main():
    number = int(input("Number: "))

    if validate(number):
        print(identify(number))
    else:
        print("INVALID")


def validate(number):
    x = str(number)
    tot1 = 0
    tot2 = 0
    for i in range(len(x)):
        if i % 2 == 0:
            temp = int(x[i]) * 2
            temp1 = 0
            if temp > 9:
                temp1 = str(temp)[0]
                temp = str(temp)[1]
            tot1 += int(temp) + int(temp1)
        else:
            tot2 += int(x[i])
    total = tot1 + tot2
    if str(total)[-1] == "0":
        return True
    else:
        return False


def identify(number):
    number = str(number)
    if len(number) == 13 or len(number) == 16 and number[0] == '4':
        return "VISA"
    if len(number) == 16 and number[0] in ['51', '52', '53', '54', '55']:
        return "MASTERCARD"
    if len(number) == 15 and number[0] == '34' or number[0] == '37':
        return "AMEX"
    else:
        return "INVALID"


main()

File: Ps.1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long long number = get_long("Number: ");
    long long card = number;
    int counter = 0;

    while (number > 0)
    {
        number /= 10;
        counter++;
    }
    if (counter != 13 && counter != 15 && counter != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    int sum1 = 0;
    int sum2 = 0;
    int mod1;
    int mod2;
    int total;
    long long rest = card;
    do
    {
        mod1 = card % 10;
        card /= 10;
        sum1 += mod1;

        mod2 = card % 10;
        card /= 10;

        if (mod2 * 2 > 9)
        {
            sum2 += (mod2 * 2) % 10;
            sum2 += (mod2 * 2) / 10;
        }
        else
        {
            mod2 *= 2;
            sum2 += mod2;
        }
    }
    while (card > 0);

    total = sum1 + sum2;
    do
    {
        rest /= 10;
    }

    while (rest > 100);

    if ((rest == 34 || rest == 37) && (counter == 15) && (total % 10 == 0))
    {
        printf("AMEX\n");
    }
    else if ((rest == 51 || rest == 52 || rest == 53 || rest == 54 || rest == 55) &&
             (counter == 16) && (total % 10 == 0))
    {
        printf("MASTERCARD\n");
    }
    else if ((rest / 10 == 4) && (counter == 13 || counter == 16) && (total % 10 == 0))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}
